blockchain/tx_pool.go

```diff
@@ -658,38 +668,42 @@ func (pool *TxPool) validateTx(tx *types.Transaction) error {
 			return ErrFeeCapVeryHigh
 		}
 
-		// TODO : Need to KIP-71 hardfork.
-		// Ensure gasFeeCap is greater than or equal to gasTipCap.
-		//if tx.GasFeeCap().Cmp(tx.GasTipCap()) < 0 {
-		//	return ErrTipAboveFeeCap
-		//}
-		//
-		//if pool.gasPrice.Cmp(tx.GasTipCap()) != 0 {
-		//	logger.Trace("fail to validate maxPriorityFeePerGas", "unitprice", pool.gasPrice, "maxPriorityFeePerGas", tx.GasFeeCap())
-		//	return ErrInvalidGasTipCap
-		//}
-		//
-		//if pool.gasPrice.Cmp(tx.GasFeeCap()) != 0 {
-		//	logger.Trace("fail to validate maxFeePerGas", "unitprice", pool.gasPrice, "maxFeePerGas", tx.GasTipCap())
-		//	return ErrInvalidGasFeeCap
-		//}
-
-		// Ensure transaction's gasFeeCap is greater than or equal to transaction pool's gasPrice(baseFee).
-		if pool.gasPrice.Cmp(tx.GasFeeCap()) > 0 {
-			logger.Trace("fail to validate maxFeePerGas", "unitPrice", pool.gasPrice, "maxFeePerGas", tx.GasFeeCap())
-			return ErrFeeCapBelowBaseFee
+		if pool.kip71 {
+			// Ensure transaction's gasFeeCap is greater than or equal to transaction pool's gasPrice(baseFee).
+			if pool.gasPrice.Cmp(tx.GasFeeCap()) > 0 {
+				logger.Trace("fail to validate maxFeePerGas", "pool.gasPrice", pool.gasPrice, "maxFeePerGas", tx.GasFeeCap())
+				return ErrFeeCapBelowBaseFee
+			}
+		} else {
+			// Ensure gasFeeCap is greater than or equal to gasTipCap.
+			if tx.GasFeeCap().Cmp(tx.GasTipCap()) < 0 {
+				return ErrTipAboveFeeCap
+			}
+
+			if pool.gasPrice.Cmp(tx.GasTipCap()) != 0 {
+				logger.Trace("fail to validate maxPriorityFeePerGas", "unitprice", pool.gasPrice, "maxPriorityFeePerGas", tx.GasFeeCap())
+				return ErrInvalidGasTipCap
+			}
+
+			if pool.gasPrice.Cmp(tx.GasFeeCap()) != 0 {
+				logger.Trace("fail to validate maxFeePerGas", "unitprice", pool.gasPrice, "maxFeePerGas", tx.GasTipCap())
+				return ErrInvalidGasFeeCap
+			}
 		}
+
 	} else {
-		// TODO : Need to KIP-71 hardfork.
-		//if pool.gasPrice.Cmp(tx.GasPrice()) != 0 {
-		//	logger.Trace("fail to validate unitprice", "unitPrice", pool.gasPrice, "txUnitPrice", tx.GasPrice())
-		//	return ErrInvalidUnitPrice
-		//}
-
-		// Ensure transaction's gasPrice is greater than or equal to transaction pool's gasPrice(baseFee).
-		if pool.gasPrice.Cmp(tx.GasPrice()) > 0 {
-			logger.Trace("fail to validate gasprice", "pool.gasPrice", pool.gasPrice, "tx.gasPrice", tx.GasPrice())
-			return ErrGasPriceBelowBaseFee
+		if pool.kip71 {
+			// Ensure transaction's gasPrice is greater than or equal to transaction pool's gasPrice(baseFee).
+			if pool.gasPrice.Cmp(tx.GasPrice()) > 0 && tx.Type() != types.TxTypeCancel {
```

Comment:
If tx is a transaction cancel type, shouldn't it be necessary to satisfy `pool.gasPrice.Cmp(tx.GasPrice()) <= 0`?
I think cancel tx should replace without comparison with gas price of old tx, but still should satisfy `pool.gasPrice.Cmp(tx.GasPrice()) <= 0`.

Reply:
I tested the case in `TestCancelTransactionAcceptedKip71`. If `tx == TxTypeCancel`, it just pass the condition. Why are you thinking `TxTypeCancel` should satisfy `pool.gasPrice.Cmp(tx.GasPrice()) <= 0`?

Reply:
If any cancelTx can be accepted, an attacker could fill the txPool with unprocessed transactions.
It should be fixed. Thanks!

Reply:
fixed it.